opt: normalize LIKE ... ESCAPE

mgartner opened the pull request:
#### logictest: make test case deterministic

This commit makes a test case in `distsql_agg` deterministic no matter
how many optimizer rules have been defined. It now explicitly disables
an optimizer rule, which is required to reproduce the behavior, rather
than using the `testing_optimizer_random_seed` and
`testing_optimizer_disable_rule_probability` session settings.

Release note: None

#### opt: make scalar list custom func accessor safer

The custom functions `FirstScalarListExpr` and `SecondScalarListExpr`
have been replaced by `ScalarExprAt`. This function is safer because it
returns ok=false if the given index is out-of-bounds instead of
panicking. This function has been added to `general_funcs.go` instead of
`comp_funcs.go` in preparation for usage outside of `comp.opt`.

Release note: None

#### opt: normalize LIKE ... ESCAPE

An expression in the form `a LIKE b ESCAPE '/'`, which is parsed as
`like_escape(a, b, '/')`, is now normalized to `a LIKE b`. This is valid
because `\` is the default escape character, so the expressions are
equivalent.

This normalization allows for further optimization, for example:

```sql
CREATE TABLE t (s STRING, INDEX (s));

EXPLAIN SELECT s FROM t WHERE s LIKE 'foo\_bar' ESCAPE '\';
-- BEFORE:
--  • filter
--  │ filter: like_escape(s, e'foo\\_bar', e'\\')
--  │
--  └── • scan
--        table: t@t_pkey
--        spans: FULL SCAN

-- AFTER:
--   • scan
--     table: t@t_s_idx
--     spans: [/'foo_bar' - /'foo_bar']
```

Informs #30192

Release note (performance improvement): Queries with filters in the form
`a LIKE b ESCAPE '\'` are now index-accelerated in some cases which they
were not before.

mgartner committed:
opt: make scalar list custom func accessor safer

The custom functions `FirstScalarListExpr` and `SecondScalarListExpr`
have been replaced by `ScalarExprAt`. This function is safer because it
returns ok=false if the given index is out-of-bounds instead of
panicking. This function has been added to `general_funcs.go` instead of
`comp_funcs.go` in preparation for usage outside of `comp.opt`.

Release note: None

mgartner committed:
opt: normalize LIKE ... ESCAPE

An expression in the form `a LIKE b ESCAPE '/'`, which is parsed as
`like_escape(a, b, '/')`, is now normalized to `a LIKE b`. This is valid
because `\` is the default escape character, so the expressions are
equivalent.

This normalization allows for further optimization, for example:

```sql
CREATE TABLE t (s STRING, INDEX (s));

EXPLAIN SELECT s FROM t WHERE s LIKE 'foo\_bar' ESCAPE '\';
-- BEFORE:
--  • filter
--  │ filter: like_escape(s, e'foo\\_bar', e'\\')
--  │
--  └── • scan
--        table: t@t_pkey
--        spans: FULL SCAN

-- AFTER:
--   • scan
--     table: t@t_s_idx
--     spans: [/'foo_bar' - /'foo_bar']
```

Informs #30192

Release note (performance improvement): Queries with filters in the form
`a LIKE b ESCAPE '\'` are now index-accelerated in some cases which they
were not before.

ZhouXing19 reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> Thanks!!

> @ZhouXing19 reviewed 3 of 3 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155064#-:-Ob4ZP2o2A8QgAgums06:b-ki7e4k)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2)



<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice!

> @michae2 reviewed 3 of 3 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155064#-:-ObEoWub0AAXBgL_tFPY:b-8gk8uj)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/opt/norm/testdata/rules/scalar` line 2719 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155064#-ObEoPw-8sovKbryGrKL:-ObEoPw0CCwu_YY9iXUT:b-pwpz93) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b83ea88a9c23ddf7cdb1ef516f1af65d2517e1d/pkg/sql/opt/norm/testdata/rules/scalar#L2719)):*
> ```
> 
> norm expect-not=ConvertLikeEscapeToLike
> SELECT s LIKE 'foo%' ESCAPE 'f' FROM a
> ```

nit: Might also be good to add a case calling `like_escape()` directly.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155064)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2 and @ZhouXing19)

___
*[`pkg/sql/opt/norm/testdata/rules/scalar` line 2719 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155064#-ObEoPw-8sovKbryGrKL:-OeI0ten9vi2fTHsjQL1:bcvjgzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b83ea88a9c23ddf7cdb1ef516f1af65d2517e1d/pkg/sql/opt/norm/testdata/rules/scalar#L2719)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

nit: Might also be good to add a case calling `like_escape()` directly.
</blockquote></details>

Good idea. Done.


<!-- Sent from Reviewable.io -->

mgartner committed:
logictest: make test case deterministic

This commit makes a test case in `distsql_agg` deterministic no matter
how many optimizer rules have been defined. It now explicitly disables
an optimizer rule, which is required to reproduce the behavior, rather
than using the `testing_optimizer_random_seed` and
`testing_optimizer_disable_rule_probability` session settings.

Release note: None
